src/azul/chalice.py

```diff
@@ -193,33 +196,211 @@ def _api_gateway_context_middleware(self, event, get_response):
         finally:
             config.lambda_is_handling_api_gateway_request = False
 
-    hsts_max_age = 60 * 60 * 24 * 365 * 2
-
-    # Headers added to every response from the app, as well as canned 4XX and
-    # 5XX responses from API Gateway. Use of these headers addresses known
-    # security vulnerabilities.
-    #
-    security_headers = {
-        'Content-Security-Policy': jw('default-src', sq('self')),
-        'Referrer-Policy': 'strict-origin-when-cross-origin',
-        'Strict-Transport-Security': jw(f'max-age={hsts_max_age};',
-                                        'includeSubDomains;',
-                                        'preload'),
-        'X-Content-Type-Options': 'nosniff',
-        'X-Frame-Options': 'DENY',
-        'X-XSS-Protection': '1; mode=block'
-    }
+    @classmethod
+    def csp_nonce(cls) -> str:
+        """
+        Return a randomly generated nonce value for use in a Content Security
+        Policy header.
+        """
+        return base64.b64encode(secrets.token_bytes(32)).decode('ascii').rstrip('=')
+
+    @classmethod
+    def content_security_policy(cls, nonce: str | None = None) -> str:
+        """
+        >>> from azul.doctests import assert_json
+        >>> assert_json(AzulChaliceApp.content_security_policy(None).split(';'))
+        [
+            "default-src 'self'",
+            "img-src 'self' data:",
+            "script-src 'self'",
+            "style-src 'self'",
+            "frame-ancestors 'none'"
+        ]
+
+        >>> assert_json(AzulChaliceApp.content_security_policy(nonce='foo').split(';'))
+        [
+            "default-src 'self'",
+            "img-src 'self' data:",
+            "script-src 'self' 'nonce-foo'",
+            "style-src 'self' 'nonce-foo'",
+            "frame-ancestors 'none'"
+        ]
+        """
+        self_ = sq('self')
+        none = sq('none')
+        nonce = [] if nonce is None else [sq('nonce-' + nonce)]
+
+        return ';'.join([
+            jw('default-src', self_),
+            jw('img-src', self_, 'data:'),
+            jw('script-src', self_, *nonce),
+            jw('style-src', self_, *nonce),
+            jw('frame-ancestors', none),
+        ])
+
+    @classmethod
+    def validate_csp(cls, csp: str, has_nonce: bool) -> None:
+        """
+        >>> cls = AzulChaliceApp
+        >>> cls.validate_csp("default-src 'self';img-src 'self' data:;"
+        ...                   "script-src 'self';style-src 'self';"
+        ...                   "frame-ancestors 'none'", has_nonce=False)
+
+        >>> cls.validate_csp("default-src 'self';img-src 'self' data:;"
+        ...                   "script-src 'self' 'nonce-1234567890123456789012345678901234567890123';"
+        ...                   "style-src 'self' 'nonce-1234567890123456789012345678901234567890123';"
+        ...                   "frame-ancestors 'none'", has_nonce=True)
+
+        # Not all expected directives are specified
```

Comment:
Remove this case. It is not one of the test cases I [specified](https://github.com/DataBiosphere/azul/pull/6483#pullrequestreview-2410415713). Remove all other cases that become redundant by asserting the expected headers (with the expected CSP injected) in the IT, as I requested in previous review.